functions/access.js

```diff
@@ -0,0 +1,123 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require("./auth");
+const { uuid } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  if (!uid) return false;
+
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const granted = await venue.ref.collection("accessgranted").doc(uid).get();
+  if (!granted.exists || !granted.tokens) {
+    return false;
+  }
+
+  if (Object.keys(granted.tokens).includes(token)) {
+    // Record that the token was checked
+    if (!granted.tokens[token].usedAt) {
+      granted.tokens[token].usedAt = [];
+    }
+    granted.tokens[token].usedAt.push(new Date().getTime());
+    await venue.ref.collection("accessgranted").doc(uid).set(granted);
+    return true;
+  }
+
+  return false;
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
```

Comment:
Is this function exposed as an API endpoint, or just used internally? If the former, might be worth us checking that `method` is on an 'allow list' rather than just being passed through directly?

Reply:
This is currently being used internally.

Reply:
Cool. So long as `method` can't be passed down from arbitrary external user input this is probably fine as is.